applegacy/src/main/java/com/hamdan/forzenbook/applegacy/viewmodels/LegacyCreateAccountViewModel.kt

```diff
@@ -0,0 +1,167 @@
+    fun test() {
+        Log.v("Hamdan", "Hello")
+    }
+
+    private var errorId: Int? = null
+    private var firstName: Entry = Entry("", LoginError.NameError.None)
+    private var lastName: Entry = Entry("", LoginError.NameError.None)
+    private var birthDay: Entry = Entry("", LoginError.BirthDateError.None)
+    private var email: Entry = Entry("", LoginError.EmailError.None)
+    private var location: Entry = Entry("", LoginError.LocationError.None)
+    private var isDateDialogOpen: Boolean = false
+    private var isLoading: Boolean = false
+
+    var onCreateAccountUpdate: ((CreateAccountState) -> Unit)? = null
+    lateinit var onAccountCreateSuccess: () -> Unit
+
+    private fun updateCreateAccountState() {
+        onCreateAccountUpdate?.invoke(
+            CreateAccountState(
+                errorId = errorId,
+                firstName = firstName,
+                lastName = lastName,
+                birthDay = birthDay,
+                email = email,
+                location = location,
+                isDateDialogOpen = isDateDialogOpen,
+                isLoading = isLoading
+            )
+        )
+    }
+
+    fun updateCreateAccountTextAndErrors(
+        inFirstName: String,
+        inLastName: String,
+        inBirthDay: String,
+        inEmail: String,
+        inLocation: String,
+    ) {
+        firstName = firstName.copy(text = inFirstName)
+        lastName = lastName.copy(text = inLastName)
+        birthDay = birthDay.copy(text = inBirthDay)
+        email = email.copy(text = inEmail)
+        location = location.copy(text = inLocation)
+        val stringStates =
+            createAccountValidationUseCase(
+                CreateAccountEntrys(
+                    firstName = firstName,
+                    lastName = lastName,
+                    birthDay = birthDay,
+                    email = email,
+                    location = location
+                )
+            )
+        firstName = stringStates.firstName
+        lastName = stringStates.lastName
+        birthDay = stringStates.birthDay
+        email = stringStates.email
+        location = stringStates.location
+        updateCreateAccountState()
+    }
+
+    fun createAccountDateDialogClicked() {
+        createAccountShowDateDialog()
+    }
+
+    private fun createAccountShowDateDialog() {
+        isDateDialogOpen = true
+        updateCreateAccountState()
+    }
+
+    fun createAccountDateDialogSubmitClicked() {
+        closeAccountShowDateDialog()
+    }
+
+    fun createAccountDateDialogDismiss() {
+        closeAccountShowDateDialog()
+    }
+
+    private fun closeAccountShowDateDialog() {
+        isDateDialogOpen = false
+        updateCreateAccountState()
+    }
+
+    private fun createAccountNormalView() {
+        errorId = null
+        updateCreateAccountState()
+    }
+
+    fun createAccountDismissErrorClicked() {
+        createAccountNormalView()
+    }
+
+    fun createAccount() {
```

Comment:
Things like this function make me think an abstract ViewModel base class is needed so there's no duplication

Reply:
I don't understand what you mean. what duplication is there in this context. Are you refering to between the two viewmodels. Are you refering to this createAccount and the createAccount in the app Viewmodel?